src/impl_views.rs

```diff
@@ -533,5 +533,108 @@ impl<'a, A, D> ArrayViewMut<'a, A, D>
+    /// let mut v1 = arr.view_mut();
+    /// let v2 = unsafe { v1.aliasing_view() };
+    /// v1[0] = 5;
+    /// assert_eq!(v2[0], 5);
+    /// ```
+    ///
+    /// Note that the underlying array is still mutably borrowed, so this won't
+    /// compile:
+    ///
+    /// ```compile_fail
+    /// // cannot borrow `arr` as immutable because it is also borrowed as mutable
+    ///
+    /// use ndarray::prelude::*;
+    ///
+    /// let mut arr = arr1(&[0, 1, 2]);
+    /// let v = unsafe { arr.view_mut().aliasing_view() };
+    /// //               --- mutable borrow occurs here
+    /// assert_eq!(arr[0], 0);
+    /// //         ^^^ immutable borrow occurs here
+    /// ```
+    ///
+    /// For the same reason, the view cannot outlive the data in `arr`, so this
+    /// won't compile:
+    ///
+    /// ```compile_fail
+    /// // `arr` does not live long enough
+    ///
+    /// use ndarray::prelude::*;
+    ///
+    /// let v = {
+    ///     let mut arr = arr1(&[0, 1, 2]);
+    ///     unsafe { arr.view_mut().aliasing_view() }
+    ///     //       --- borrow occurs here
+    /// };
+    /// // `arr` dropped here while still borrowed
+    /// ```
+    #[doc(hidden)]
+    pub unsafe fn aliasing_view(&self) -> ArrayView<'a, A, D> {
+        ArrayView::new_(self.ptr, self.dim.clone(), self.strides.clone())
+    }
+
+    /// Returns a new `ArrayViewMut` referencing the same data.
+    ///
+    /// This is unsafe because it produces two mutable borrows of the same
+    /// data.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use ndarray::prelude::*;
+    ///
+    /// let mut arr = arr1(&[0, 1, 2]);
+    /// let mut v1 = arr.view_mut();
+    /// let v2 = unsafe { v1.aliasing_view_mut() };
+    /// v1[0] = 5;
+    /// assert_eq!(v2[0], 5);
+    /// ```
+    ///
+    /// Note that the underlying array is still mutably borrowed, so this won't
+    /// compile:
+    ///
+    /// ```compile_fail
+    /// // cannot borrow `arr` as immutable because it is also borrowed as mutable
+    ///
+    /// use ndarray::prelude::*;
+    ///
+    /// let mut arr = arr1(&[0, 1, 2]);
+    /// let v = unsafe { arr.view_mut().aliasing_view_mut() };
+    /// //               --- mutable borrow occurs here
+    /// assert_eq!(arr[0], 0);
+    /// //         ^^^ immutable borrow occurs here
+    /// ```
+    ///
+    /// For the same reason, the view cannot outlive the data in `arr`, so this
+    /// won't compile:
+    ///
+    /// ```compile_fail
+    /// // `arr` does not live long enough
+    ///
+    /// use ndarray::prelude::*;
+    ///
+    /// let v = {
+    ///     let mut arr = arr1(&[0, 1, 2]);
+    ///     unsafe { arr.view_mut().aliasing_view_mut() }
+    ///     //       --- borrow occurs here
+    /// };
+    /// // `arr` dropped here while still borrowed
+    /// ```
+    #[doc(hidden)]
+    pub unsafe fn aliasing_view_mut(&self) -> ArrayViewMut<'a, A, D> {
+        ArrayViewMut::new_(self.ptr, self.dim.clone(), self.strides.clone())
```

Comment:
Here I think we must choose a different design. It's tricky, and connected with the guidelines for unsafe code and the exact safety rules we have to follow, which is not entirely set in stone yet.

The gist of it is that I think `ArrayViewMut<'a, T, D>` has the same semantic as `&'a mut T`, a mutable reference.

Even in unsafe code we are not allowed to create multiple mutable pointers to the same value and use them as the type `&mut T`, so I think this method crosses that line. What we do in usual unsafe code is to keep raw pointers.

The reasoning isn't exactly bullet proof since it relies on the incompleteness of the unsafe code guidelines; so all the more reason to be careful.

I haven't understood the whole code here yet, but the ndarray equivalent would be something like `type RawView<T, D> = ArrayBase<PhantomData<*mut T>, D>`

Reply:
Yeah, getting this right is tricky. We need to ensure that (1) the array being sliced is mutably borrowed and (2) the views have the correct lifetime. Applying Rust's pointer aliasing rules to (temporary) `ArrayView`/`ArrayViewMut` instances complicates things further.

If it matters, `.aliasing_view_mut(&self)` is hidden from the docs, and it's not possible to create aliasing mutable views with `multislice!()`. (It panics if you try to do so.) However, it sounds like you're saying that the implementation should never create (even temporary, internal) aliasing `ArrayViewMut` instances.

At first glance, I see a couple of options. Both options include the following:

* Add a method `.into_repr(self) -> ViewRepr<&'a mut A>` to `ArrayViewMut<'a, A, D>`.
* Add a method `.raw_strides(&self)` to `ArrayBase` similar to `.raw_dim(&self)` but for strides.
* Add an associated function `from_raw_parts(data: ViewRepr<&'a mut A>, ptr: *mut A, dim: D, strides: D)` to `ArrayViewMut<'a, A, D>`.
* Remove the `A: Copy` constraint on the `Copy` implementation for `ViewRepr<A>`. This should be okay, right? (We should also remove the `A: Clone` constraint on the `Clone` implementation for consistency.)

#### Option 1

* Refactor the body of `.slice()` (and as a result, `.slice_move()`, `.slice_inplace()`, `.slice_axis_inplace()`, and `.subview_inplace()`) into a function `slice()` that operates on `ptr: *mut S::Elem, dim: &D, strides: &D, info: &SliceInfo<D::SliceArg, Do>` instead of operating on `ArrayBase` and `info`.
* Then, `multislice!(arr, (mut info1, mut info2))` would expand to something like

  ```rust
  let view = arr.view_mut();
  let ptr = view.as_mut_ptr();
  let dim = view.raw_dim();
  let strides = view.raw_strides();
  let repr = view.into_repr();
  (
      {
          let (new_ptr, new_dim, new_strides) = slice(ptr, &dim, &strides, info1);
          unsafe { ArrayViewMut::from_raw_parts(repr, new_ptr, new_dim, new_strides) }
      },
      {
          let (new_ptr, new_dim, new_strides) = slice(ptr, &dim, &strides, info2);
          unsafe { ArrayViewMut::from_raw_parts(repr, new_ptr, new_dim, new_strides) }
      },
  )
  ```

  plus checks that the views are non-aliasing.

#### Option 2

* Remove the `S: Data` constraint on `ArrayBase`, `.slice_move()`, `.slice_inplace()`, `.slice_axis_inplace()`, `.subview_inplace()`, `.ndim()`, and `.len_of()`. (We can't have `PhantomData<*mut A>` implement `Data` because that would allow dereferencing essentially a raw pointer in safe code (e.g. with `.index()`).)
* Add a method `.as_mut_ptr_view(&self) -> ArrayBase<PhantomData<*mut A>, D>` to `ArrayViewMut<'a, A, D>`.
* Then, `multislice!(arr, (mut info1, mut info2))` would expand to something like

  ```rust
  let view = arr.view_mut();
  let ptr_view = view.as_mut_ptr_view();
  let repr = view.into_repr();
  (
      {
          let sliced = ptr_view.clone().slice_move(info1);
          unsafe {
              ArrayViewMut::from_raw_parts(
                  repr,
                  sliced.as_mut_ptr(),
                  sliced.raw_dim(),
                  sliced.raw_strides(),
              )
          }
      },
      {
          let sliced = ptr_view.clone().slice_move(info2);
          unsafe {
              ArrayViewMut::from_raw_parts(
                  repr,
                  sliced.as_mut_ptr(),
                  sliced.raw_dim(),
                  sliced.raw_strides(),
              )
          }
      },
  )
  ```

  plus checks that the views are non-aliasing.

What do you think of these approaches?

Reply:
- IMO it doesn't matter that `aliasing_view_mut(&self)` is hidden from docs. The `&X -> &mut Y` signature is just inadmissible.
- The  the A: Copy constraint for a ViewRepr is important, without it, `ArrayViewMut` can be `Copy`.
- I think we will find that “Remove the S: Data constraint on ArrayBase” is not possible - we need something to give us the type of `*const S::Data`.
- FYI `ArrayViewMut`

I think we can just deconstruct the view into raw parts and construct each `ArrayViewMut` in turn. The main concern seems to be about how to reach the `slice_move` method?

Basically Option 2, but I wouldn't want to change any of the methods in the way you suggest. I know this answer is not complete.

Feel free to ping me when I forget to answer.

Reply:
> the A: Copy constraint for a ViewRepr is important, without it, `ArrayViewMut` can be `Copy`.

I'm confused about this, because the constraints on the `Copy` implementation of `ArrayBase` are `S: DataClone + Copy, D: Copy`, and `DataClone` is not implemented for `ViewRepr<&'a mut A>`. As a result, it seems like the `A: Copy` constraint shouldn't influence whether `ArrayViewMut` is `Copy`. Am I missing something?

Would it be okay to remove the `A: Clone` constraint for the `Clone` implementation for `ViewRepr<A>`?

Regardless, neither of these constraint changes are critical; there are alternatives.

> “Remove the S: Data constraint on ArrayBase” is not possible - we need something to give us the type of `*const S::Data`.

Good point. (I assume you meant `S::Elem`.) I wouldn't want to have to add a third type parameter.

> The main concern seems to be about how to reach the `slice_move` method?

Yeah, that's the issue. We have to either refactor the slicing methods so that we can slice the deconstructed pieces (`ptr`, `dim`, `strides`) or create (temporarily) aliasing `ArrayBase` instances in order to call `slice_move`. We can't slice `ArrayViewMut` views of the full array because that would require multiple (temporary) aliasing mutable views.

Here are a couple more options to avoid refactoring the slicing methods:

#### Option 3

Add a (hidden, unsafe) method to convert `ArrayView` to `ArrayViewMut`. Then, `multislice!()` could slice the array into `ArrayView` instances, and then "upgrade" them to `ArrayViewMut` instances. I don't really like this idea, though, because it seems weird to have a method that effectively turns `&T` into `&mut T`. It would also take a little work to make sure the underlying array is mutably borrowed.

#### Option 4

Add a new marker trait `pub unsafe trait DataDeref: Data {}` that would be required to read/write data (i.e. deref `self.ptr`) in an `ArrayBase` with safe methods. `DataDeref` would be implemented for all of the existing data types, but would not be implemented for `PhantomData<*mut A>`. This would allow us to safely implement `Data` for `PhantomData<*mut A>` so that you could create an `ArrayBase<PhantomData<*mut A>, D>`.

With this approach, `multislice!()` could create "pointer views" (`ArrayBase<PhantomData<*mut A>, D>`) of the entire array, slice them, and then "dereference" them (or use `from_raw_parts()`) to get the final `ArrayViewMut` instances.

This seems like a fairly nice, clean approach, but it is a breaking change because the constraint on many `ArrayBase` methods (e.g. `.index()` and `.scalar_sum()`) would change from `S: Data` to `S: DataDeref`.

What do you think?

Reply:
Option 3 is not an alternative for the same reason that we want to fix this method to start with.

I'd go with Option 4, except making Data require a new trait instead, `DataRaw` or something like that. It shouldn't be breaking then. It would be very interesting to go in this direction anyway -- a raw pointer ArrayBase with no specific ownership semantic

Reply:
You are right about the Copy thing, I didn't think about that.

Reply:
Yeah, I like the idea to create a "raw pointer array" too. I'm not sure I understand what you're proposing with `DataRaw`, though. In order to be able to create an `ArrayBase<PhantomData<*mut A>, D>`, `PhantomData<*mut A>` must implement `Data`. However, it's not safe to read/write data from this type because that's equivalent to dereferencing the pointer. So, we need another trait to mean "you can safely read/write the data" that we can use as a bound for methods that read/write data. We can't make `Data` require this new trait, though, because then `PhantomData<*mut A>` can't implement `Data`.

Reply:
Yes you are down the same thread of what I mean.